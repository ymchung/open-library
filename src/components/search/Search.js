import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { useNavigate } from 'react-router-dom';
import classNames from 'classnames';

const namespace = 'search';

const Search = ({ fetchBooks, searchText }) => {
    const [inputValue, setInputValue] = useState('');
    const [isValidInput, setIsValidInput] = useState(true);

    const navigate = useNavigate();

    useEffect(() => {
        if (searchText !== null) {
            setInputValue(searchText);
        }
        return () => {
            setIsValidInput(true);
        }
    }, [searchText])

    const onInputChange = e => {
        setInputValue(e.target.value);
    };

    const handleOnSubmit = e => {
        e.preventDefault();
        
        const value = inputValue.trim();
        if (value !== '' && value.length > 3) {
            const queryString = value.replace(/ /g,'+');
            setIsValidInput(true);
            fetchBooks(value);
            navigate(`/books?q=${queryString}`);
        } else {
            setIsValidInput(false);
        }
    };

    return (
        <form className={namespace}>
            <div className={`${namespace}__container input-group`}>
                <input type="text" className={classNames(`${namespace}__input`, 'form-control', {'is-invalid': !isValidInput})} placeholder="Buscar libro" aria-label="" value={inputValue} onChange={onInputChange} />
                <button type="submit" className={`${namespace}__button btn`} onClick={handleOnSubmit}>
                    <span>BUSCAR</span>
                </button>
                {!isValidInput && <div className="invalid-feedback">Ingrese m√°s de 3 caracteres</div>}
            </div>
        </form>
    );
};

Search.propTypes = {
    searchText: PropTypes.string,
    fetchBooks: PropTypes.func.isRequired,
};

Search.defaultProps = {
    searchText: '',
};

export default Search;