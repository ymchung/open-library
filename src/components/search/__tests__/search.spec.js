import { render, fireEvent } from '@testing-library/react';
import Search from '../Search';

jest.mock("react-router-dom", () => ({
    useNavigate: () => jest.fn(),
  }));

describe('Search Component', () => {
    const props = {
        searchText: 'abcd',
        fetchBooks: jest.fn(),
    }

    test('Component should render correctly', () => {
        const { container } = render(<Search {...props} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    test('On search should call fetchBooks function', () => {
        const { getByTestId } = render(<Search {...props} />);
        fireEvent.click(getByTestId('search__button'));

        expect(props.fetchBooks).toHaveBeenCalledTimes(1);
    });
});