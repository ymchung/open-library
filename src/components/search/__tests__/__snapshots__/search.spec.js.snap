// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component Component should render correctly 1`] = `
<form
  class="search"
>
  <div
    class="search__container input-group"
  >
    <input
      aria-label="Buscador"
      class="search__input form-control"
      placeholder="Buscar libro"
      type="text"
      value="abcd"
    />
    <button
      class="search__button btn"
      data-testid="search__button"
      type="submit"
    >
      <span>
        BUSCAR
      </span>
    </button>
  </div>
</form>
`;
